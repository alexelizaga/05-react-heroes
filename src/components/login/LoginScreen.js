import { useContext } from 'react';
import { useNavigate } from 'react-router-dom';

import { AuthContext } from '../../auth/authContext';
import { types } from '../../types/types';

export const LoginScreen = () => {
  const { dispatch } = useContext(AuthContext);
  const navigate = useNavigate();

  const handleLogin = () => {
    const action = {
      type: types.login,
      payload: { name: 'Alex' }
    }

    dispatch(action);

    const lastPath = localStorage.getItem('lastPath') || '/marvel';

    navigate(lastPath, {
      replace: true,
    });

  }

  return (
    <div className="container mt-5">
      <h1 data-testid='LoginScreenTitle'>Login</h1>
      <hr />

      <button
        data-testid='LoginScreenBtn'
        className="btn btn-primary"
        onClick={handleLogin}
      >
        Login
      </button>
    </div>
  )
}
