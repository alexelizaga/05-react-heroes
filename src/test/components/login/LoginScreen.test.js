import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';

import { AuthContext } from '../../../auth/authContext';
import { LoginScreen } from '../../../components/login/LoginScreen';
import { types } from '../../../types/types';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}));

describe('Test <LoginScreen />', () => {

  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: false
    }
  };

  test('should render correctly', () => {
    const view = render(
      <AuthContext.Provider value={ contextValue }>
        <MemoryRouter initialEntries={['/login']}>
          <LoginScreen />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    expect(view.baseElement).toMatchSnapshot();
  });

  test('should login', () => {
    render(
      <AuthContext.Provider value={ contextValue }>
        <MemoryRouter initialEntries={['/login']}>
          <Routes>
            <Route path='/login' element={<LoginScreen />} />
          </Routes>
        </MemoryRouter>
      </AuthContext.Provider>
    );

    const loginScreenBtn = screen.queryByTestId('LoginScreenBtn');
    fireEvent.click(loginScreenBtn);

    expect(contextValue.dispatch).toBeCalledWith({
      type: types.login,
      payload: { name: 'Alex'} 
    });

    expect(mockNavigate).toBeCalledWith('/marvel', { replace: true });

    localStorage.setItem('lastPath', '/dc');

    fireEvent.click(loginScreenBtn);

    expect(mockNavigate).toBeCalledWith('/dc', { replace: true });
  });

});