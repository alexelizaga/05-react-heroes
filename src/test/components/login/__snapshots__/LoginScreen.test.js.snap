// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <LoginScreen /> should render correctly 1`] = `
<body>
  <div>
    <div
      class="container mt-5"
    >
      <h1
        data-testid="LoginScreenTitle"
      >
        Login
      </h1>
      <hr />
      <button
        class="btn btn-primary"
        data-testid="LoginScreenBtn"
      >
        Login
      </button>
    </div>
  </div>
</body>
`;
